src/lightning/fabric/utilities/distributed.py

```diff
@@ -23,9 +28,66 @@ class group:  # type: ignore
         WORLD = None
 
 
+if TYPE_CHECKING:
+    from lightning.fabric.strategies import Strategy
+
+
 log = logging.getLogger(__name__)
 
 
+@lru_cache(maxsize=128)
+def is_shared_filesystem(strategy: "Strategy", path: Optional[_PATH] = None, timeout: int = 3) -> bool:
+    """Checks whether the filesystem under the given path is shared across all processes.
+
+    Args:
+        strategy: The strategy being used, either from Fabric (``fabric.strategy``) or from Trainer
+            (``trainer.strategy``).
+        path: The path to check. Defaults to the current working directory. The user must have permissions to write
+            to this path or the parent folder, and the filesystem must be writable.
+        timeout: If any of the processes can't list the file created by rank 0 within this many seconds, the
+            filesystem is determined to be not shared.
+
+    """
+    # Fast path: Any non-local filesystem is considered shared (e.g., S3)
+    if path is not None and fsspec.utils.get_protocol(str(path)) != "file":
+        return True
+
+    path = Path(Path.cwd() if path is None else path).resolve()
+
+    # Fast path: Only distributed strategies can detect shared filesystems
+    if not hasattr(strategy, "world_size") or strategy.world_size == 1:
```

Comment:
```suggestion
    if not getattr(strategy, "world_size", 0) == 1:
```

Reply:
This isn't equivalent to the previous statement. If `strategy.world_size = 1` then we have `not 1 == 1 ` and this is False. But we want True (to exit early). So I need to keep it like that.